gpflow/monitor.py

```diff
@@ -0,0 +1,245 @@
+
+    def __init__(self, log_dir: str, callback: Callable[[], float], name: str, period: int = 1):
+        """
+        :param log_dir: directory in which to store the tensorboard files.
+            For example, './logs/my_run/'.
+        :param callback: callback to be executed and result written to TensorBoard.
+            A callback can have arguments (e.g. data) passed to the function using
+            keyword arguments.
+            For example:
+            ```
+            lambda cb(x=None): 2 * x
+            task = ScalarToTensorBoard(logdir, cb, "callback")
+
+            # specify the argument of the function using kwargs, the names need to match.
+            task(step, x=1)
+            ```
+        :param name: name used in TensorBoard.
+        :param period: defines how often to run the task; it will execute every `period`th step.
+            For large values of `period` the task will be less frequently run.
+        """
+        super().__init__(log_dir, period)
+        self.name = name
+        self.callback = callback
+
+    def run(self, **kwargs):
+        tf.summary.scalar(self.name, self.callback(**kwargs), step=self.current_step)
+
+
+class ImageToTensorBoard(ToTensorBoard):
+    def __init__(
+        self,
+        log_dir: str,
+        plotting_function: Callable[[Figure, Axes], Figure],
+        name: Optional[str] = None,
+        period: int = 1,
+        *,
+        fig_kw: Optional[Dict[str, Any]] = None,
+        subplots_kw: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        :param log_dir: directory in which to store the tensorboard files.
+            Can be a nested: for example, './logs/my_run/'.
+        :param plotting_function: function performing the plotting.
+        :param name: name used in TensorBoard.
+        :param period: defines how often to run the task; it will execute every `period`th step.
+            For large values of `period` the task will be less frequently run.
+        :params fig_kw: Keywords to be passed to Figure constructor, such as `figsize`.
+        :params subplots_kw: Keywords to be passed to figure.subplots constructor, such as
+            `nrows`, `ncols`, `sharex`, `sharey`. By default the default values 
+            from matplotlib.pyplot as used.
+        """
+        super().__init__(log_dir, period)
+        self.plotting_function = plotting_function
+        self.name = name
+        self.file_writer = tf.summary.create_file_writer(log_dir)
+        self.name = name
+        self.fig_kw = fig_kw or {}
+        self.subplots_kw = subplots_kw or {}
+
+        self.fig = Figure(**self.fig_kw)
+        if self.subplots_kw != {}:
+            self.axes = self.fig.subplots(**self.subplots_kw)
+        else:
+            self.axes = self.fig.add_subplot(111)
+
+    def _clear_axes(self):
+        try:
+            for ax in self.axes.flatten():
+                ax.clear()
+        except (AttributeError, TypeError):
+            self.axes.clear()
```

Comment:
nit/thought (happy to keep it as is, too):
```suggestion
        if isinstance(self.axes, np.ndarray):
            for ax in self.axes.flatten():
                ax.clear()
        else:
            self.axes.clear()
```
?